java/arcs/core/data/expression/TypeEvaluator.kt

```diff
@@ -0,0 +1,155 @@
+    }).apply {
+      @Suppress("SENSELESS_COMPARISON") if (this == null && !expr.nullSafe) {
+        throw IllegalArgumentException("Field '${expr.field}' not looked up on null scope")
+      }
+    }?.lookup(expr.field) as InferredType
+
+  override fun <E> visit(expr: Expression.QueryParameterExpression<E>, ctx: Scope): InferredType {
+    return parameterScope.lookup(expr.paramIdentifier) as? InferredType
+      ?: throw IllegalArgumentException(
+        "Unbound parameter '${expr.paramIdentifier}'"
+      )
+  }
+
+  override fun visit(expr: Expression.NumberLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.TextLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.BooleanLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.NullLiteralExpression, ctx: Scope) =
+    InferredType.Primitive.NullType
+
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
+
+  private fun InferredType.asScope() = (this as SeqType).type as ScopeType
+
+  override fun visit(expr: Expression.FromExpression, ctx: Scope): InferredType {
+    val scope = seqToScopeOr(expr.qualifier?.accept(this, ctx), ctx)
+
+    val resultSeq = expr.source.accept(this, scope) as SeqType
+    return SeqType(ScopeType(scope.builder().set(expr.iterationVar, resultSeq.type).build()))
+  }
+
+  override fun visit(expr: Expression.WhereExpression, ctx: Scope): InferredType {
+    return expr.qualifier.accept(this, ctx)
+  }
+
+  override fun visit(expr: Expression.LetExpression, ctx: Scope): InferredType {
+    val scopeType = expr.qualifier.accept(this, ctx).asScope()
+
+    return SeqType(
+      ScopeType(
+        scopeType.scope.builder().set(
+          expr.variableName,
+          expr.variableExpr.accept(this, scopeType.scope)
+        ).build()
+      )
+    )
+  }
+
+  override fun <T> visit(expr: Expression.SelectExpression<T>, ctx: Scope): InferredType {
+    return SeqType(
+      expr.expr.accept(
+        this,
+        expr.qualifier.accept(this, ctx).asScope().scope
+      )
+    )
+  }
+
+  override fun visit(expr: Expression.NewExpression, ctx: Scope): InferredType {
+    val initScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
+    return ScopeType(
+      expr.fields.fold(initScope.builder()) { builder, (fieldName, fieldExpr) ->
+        builder.set(fieldName, fieldExpr.accept(this, ctx))
+      }.build()
+    )
+  }
+
+  override fun <T> visit(expr: Expression.FunctionExpression<T>, ctx: Scope): InferredType {
+    val arguments = expr.arguments.map { it.accept(this, ctx) }.toList()
+    return expr.function.inferredType(this, arguments)
+  }
+
+  override fun <T> visit(expr: Expression.OrderByExpression<T>, ctx: Scope): InferredType {
+    return expr.qualifier.accept(this, ctx)
```

Comment:
verify that expressions by which we order are... I guess any primitive is allowed, but scopes and sequences are not?

Reply:
Yes, primitives are allowed. Union is allowed if and only if the types are all primitives. So really, only seqtype and scopetype are banned.